CASSJAVA-80: Support configuration to disable DNS reverse-lookups for SAN validation

aratno opened the pull request.

frankgh commented on core/src/main/java/com/datastax/oss/driver/api/core/config/DefaultDriverOption.java:
```diff
@@ -243,6 +243,12 @@ public enum DefaultDriverOption implements DriverOption {
+   *
+   * <p>Value-type: boolean
+   */
+  SSL_DNS_REVERSE_LOOKUP_SAN("advanced.ssl-engine-factory.dns-reverse-lookup-san"),
```
my only concern with this property is that the reverse lookup is not guaranteed. So the name does not seem to truly capture the behavior. From javadocs:

> This method may trigger a name service reverse lookup if the address was created with a literal IP address.

emphasize on __may__

frankgh reviewed (commented):
Looks good in general. Do we need to do the same for [ProgrammaticSslEngineFactory](https://github.com/apache/cassandra-java-driver/blob/4.x/core/src/main/java/com/datastax/oss/driver/api/core/ssl/ProgrammaticSslEngineFactory.java) and [SniSslEngineFactory](https://github.com/apache/cassandra-java-driver/blob/4.x/core/src/main/java/com/datastax/oss/driver/internal/core/ssl/SniSslEngineFactory.java)?

aratno committed:
CASSJAVA-80: Support configuration to disable DNS reverse-lookups for SAN validation

patch by Abe Ratnofsky; reviewed by Alexandre Dutra, Andy Tolbert, and Francisco Guerrero for CASSJAVA-80

adutra reviewed (approved):
LGTM, I vaguely remember a similar issue fixed in b2b9bee9152fecfdbefda654ccf578325e2a36d8, so this change makes sense.

tolbertam commented:
> Looks good in general. Do we need to do the same for [ProgrammaticSslEngineFactory](https://github.com/apache/cassandra-java-driver/blob/4.x/core/src/main/java/com/datastax/oss/driver/api/core/ssl/ProgrammaticSslEngineFactory.java) and [SniSslEngineFactory](https://github.com/apache/cassandra-java-driver/blob/4.x/core/src/main/java/com/datastax/oss/driver/internal/core/ssl/SniSslEngineFactory.java)?

Definitely see the value in doing it for the Programmatic one (I see you already covered that, thanks!), for SNI I think there is less utility as you are likely using a DNS name with the IP of the node to target in the SNI, but I think we might as well add it though just in the event that someone is using a literal IP address for the primary endpoint for tunneling through SNI.

tolbertam reviewed (approved):
:+1: looks great;  If you don't mind also adding this for the SNI factory for completeness that'd be good, but i'm +1 either way.

tolbertam commented on core/src/main/java/com/datastax/oss/driver/internal/core/ssl/SniSslEngineFactory.java:
```diff
@@ -38,9 +38,16 @@ public class SniSslEngineFactory implements SslEngineFactory {
 
   public SniSslEngineFactory(SSLContext sslContext) {
     this.sslContext = sslContext;
+    this.allowDnsReverseLookupSan = true;
```
tiny nit: can we change this constructor to `this(sslContext, true);`

tolbertam commented on core/src/main/java/com/datastax/oss/driver/internal/core/ssl/SniSslEngineFactory.java:
```diff
@@ -38,9 +38,15 @@ public class SniSslEngineFactory implements SslEngineFactory {
+    this(sslContext, true);
+  }
+
+  public SniSslEngineFactory(SSLContext sslContext, boolean allowDnsReverseLookupSan) {
```
Didn't dawn on me at the time that `SniSslEngineFactory` does not access `DriverContext` so we can't make use of it unless programatically (outside of `CloudConfigFactory`), but I suppose if someone wants to use it separately, they can programmatically and in which case this adds some functionality :+1:

frankgh reviewed (approved):
+1 thanks for the fix
